### ISSUE TITLE
Consider using stderr for Event and Debug messages

### ISSUE THREAD

**Issue by istathar** [aesiniath/unbeliever#48]
Went to a _lot_ of effort to unify stderr and stdout. Really, though, this is about one stream not trampling on the other and recognizing that in the containerized use case output streams are de-facto combined into one by whatever logging service you're using (and even earlier, at container output level).

For command-line programs, however, there is often a need to report verbose status while still passing data on stdout to a Shell pipeline. But right now the file written to by
```
 ./program --verbose > output.json
```
will be full of a lot of cruft.

It turns out that we could quite easily restore using both fd 1 an fd 2; the key was to not intermix them, so we could conceivably issue a `flush` after each line. That would be "slower" but would maintain correctness of writing ordered messages to the console.

**Comment by carlosdagos** [comment]
I wouldn't consider Event or Debug messages for stderr. My rationale is that both types of messages are typically used for analysing application behaviour. 

For example, we will typically do `./program arg1 --opt1 | grep 'some relevant string' | awk '{print $whatever}' | sort | uniq -c` or something like that (my example is actually mostly taken from my current terminal history).

There's two reasons why I would want Event and Debug in stdout:
1. When I analyse program output, I specifically analyse stdout
2. Anything coming from stderr should be exceptional (i.e. unexpected) logging.

In this case I mean _unexpected_ as in explicitly uncaught. So basically whatever is caught here: https://github.com/aesiniath/unbeliever/blob/master/core-program/lib/Core/Program/Execute.hs#L128-L160

Now with that said, `bail` does:

```haskell
setVerbosityLevel Debug
event text
```

Which, if that went to `stderr`, then it would comply with `2` above. But I actually think that code should be different.

So to sum up:
* Keep Event and Debug messages in stdout.
* Uncaught exceptions should go to stderr.
* Perhaps add more functionality so that library consumers can explicitly redirect specific logs to stderr.

Hope this helps :smile:

**Comment by istathar** [comment]
@carlosdagos I agree with your rationale, and that's why I picked the current behaviour in the first place; `2>&1` is tedious.

I did vaguely speculate that the (stubbed) _logging channel_ would be for Event (and Debug) output; perhaps I can put an option sending "logs" to stderr as one of the permitted variations that I was musing about in https://github.com/aesiniath/unbeliever/pull/13.

Not sure whether implementing uncaught exceptions to stderr is a thing; right now we've gone to a bit of trouble to capture and report them as Event level messages before terminating. Something to think about.
